pest/src/prec_climber.rs

```diff
@@ -79,137 +91,322 @@ impl<R: RuleType> BitOr for Operator<R> {
+/// be mapped.
+///
+/// [`map_primary`]: struct.PrecClimber.html#method.map_primary
+/// [`PrecClimber`]: struct.PrecClimber.html 
+pub struct PrecClimberMap<'climber, 'pest, R, F, T>
     where
-        P: Iterator<Item = Pair<'i, R>>,
-        F: FnMut(Pair<'i, R>) -> T,
-        G: FnMut(T, Pair<'i, R>, T) -> T
+    R: RuleType,
+    F: Fn(Pair<'pest, R>) -> T,
+    R: 'climber,
+{
+    ops: &'climber HashMap<R, (OpKind, Prec)>,
+    primary: F,
+    prefix: Option<PrefixFn<'pest, R, T>>,
+    suffix: Option<SuffixFn<'pest, R, T>>,
+    infix: Option<InfixFn<'pest, R, T>>,
+    phantom: PhantomData<T>,
+}
+
+impl<'climber, 'pest, R, F, T> PrecClimberMap<'climber, 'pest, R, F, T>
+    where
+    R: RuleType,
+    F: Fn(Pair<'pest, R>) -> T,
+{
+
+    /// Maps prefix operators with closure `prefix`.
+    pub fn map_prefix<X>(mut self, prefix: X) -> Self
+        where X: Fn(Pair<'pest, R>, T) -> T + 'pest
+    {
+        self.prefix = Some(Box::new(prefix));
+        self
+    }
+
+    /// Maps suffix operators with closure `suffix`.
+    pub fn map_suffix<X>(mut self, suffix: X) -> Self
+        where X: Fn(T, Pair<'pest, R>) -> T + 'pest 
     {
-        let lhs = primary(
-            pairs
-                .next()
-                .expect("precedence climbing requires a non-empty Pairs")
-        );
-        self.climb_rec(lhs, 0, &mut pairs.peekable(), &mut primary, &mut infix)
+        self.suffix = Some(Box::new(suffix));
+        self
+    }
+
+    /// Maps infix operators with a closure `infix`.
+    pub fn map_infix<X>(mut self, infix: X) -> Self
+        where X: Fn(T, Pair<'pest, R>, T) -> T + 'pest
+    {
+        self.infix = Some(Box::new(infix));
+        self
     }
 
-    fn climb_rec<'i, P, F, G, T>(
+    // Debug only: Verify definition of PrecClimberMap
+    #[cfg(debug_assertions)]
+    fn verify(&self) {
+        let mut has_prefix = false;
+        let mut has_suffix = false;
+        let mut has_infix = false;
+        self.ops.values().for_each(|(kind, _)| match kind {
+            OpKind::Prefix   => has_prefix = true,
+            OpKind::Suffix   => has_suffix = true,
+            OpKind::Infix(_) => has_infix  = true,
+        });
+        let check = |has_op: bool, has_closure: bool, name: &str| {
+            if has_op && !has_closure {
+              panic!("Climb aborted! \
+                     Found `.op(Op::{}(...))`, but no `.map_{}(...)`", name, name);
+            } else if !has_op && has_closure {
+              panic!("Climb aborted! \
+                     Found `.map_{}(...)`, but no `.op(Op::{}(...))`", name, name);
+            }
+        };
+        check(has_prefix, self.prefix.is_some(), "prefix");
+        check(has_suffix, self.suffix.is_some(), "suffix");
+        check(has_infix,  self.infix.is_some(),  "infix");
+    }
+
+    /// Start climbing `pairs`.
+    pub fn climb<P: Iterator<Item = Pair<'pest, R>>>(
+        &self,
+        pairs: P
+    ) -> T {
+
+        #[cfg(debug_assertions)]
```

Comment:
Maybe it would be worthwhile to introduce a third type that this one would get compiled to. This would both offer a nice API, i.e. return a `None` if missing `map_*`s, and relive the user from possible headaches in release builds.

Reply:
I changed so that the code panics either when:
* The stream of tokens does not match the expected pattern.
* No map function is specified for a certain kind of operator.